### PR TITLE
[3.3.5] Combat/Threat rewrite - prep & refactor PR

### PR DESCRIPTION
**Pull request by Treeston**
Add a compatibility layer to ThreatManager and HostileReference that allows scripts to be migrated. Also, various renames and refactors.

Expedited review appreciated (most of the changes are purely renames or 1:1 replacements) so we can cut down the diff set of #19930.

### CODE REVIEW

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp [hunk 1]
```diff
@@ -200,6 +200,47 @@ void ScriptedAI::DoPlaySoundToSet(WorldObject* source, uint32 soundId)
     source->PlayDirectSound(soundId);
 }
 
+void ScriptedAI::AddThreat(Unit* victim, float amount, Unit* who)
+{
+    if (!who)
+        who = me;
+    who->AddThreat(victim, amount, nullptr, true, true);
```
Comment (Shauren) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: Would like some consistency here, some of these functions are on Unit (AddThreat, ResetThreatList) and some are on ThreatManager (ModifyThreatByPercent, ResetThreat, GetThreat)

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -20285,14 +20285,36 @@ void Player::UpdateAfkReport(time_t currTime)
...
+    if (attackedPlayer && (attackedPlayer == this || (duel && duel->opponent == attackedPlayer)))
+        return;
+
+    SetContestedPvPTimer(30000);
+    if (!HasUnitState(UNIT_STATE_ATTACK_PLAYER))
+    {
+        AddUnitState(UNIT_STATE_ATTACK_PLAYER);
+        SetFlag(PLAYER_FLAGS, PLAYER_FLAGS_CONTESTED_PVP);
+        // call MoveInLineOfSight for nearby contested guards
+        UpdateObjectVisibility();
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: This is a very expensive call, if all you need is calling the AI hook then use Trinity::AIRelocationNotifier directly

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (Treeston) [inline src/server/game/Entities/Player/Player.cpp]: Sure, I just copied the existing code - moved from `Unit` to `Player` since it's only used for players.

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp (hunk 1 above)
Comment (Treeston) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: `AddThreat` is on `Unit` here because `AddThreat` was on `Unit` before. In the full PR, the method on `Unit` is merely [a forward to the threat manager](https://github.com/Treeston/TrinityCore/blob/364fe8af7618a9aad88da6e68618dc652513f032/src/server/game/Entities/Unit/Unit.cpp#L9360-L9364).

Comment (Shauren) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: If they dont do anything on Unit then why are they there

Comment (Treeston) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: Convenience, mostly. I suppose I can just as easily refactor them out.

File: src/server/game/Combat/ThreatManager.cpp [hunk 3]
```diff
@@ -391,6 +394,30 @@ HostileReference* ThreatContainer::selectNextVictim(Creature* attacker, HostileR
...
+    GetOwner()->getHostileRefManager().threatAssist(victim, amount, spell);
+}
+
+void ThreatManager::AddThreat(Unit* victim, float amount, SpellInfo const* spell, bool ignoreModifiers, bool ignoreRedirection)
+{
+    (void)ignoreModifiers; (void)ignoreRedirection;
+    if (!iOwner->CanHaveThreatList() || iOwner->HasUnitState(UNIT_STATE_EVADE))
+        return;
+    iOwner->SetInCombatWith(victim);
+    victim->SetInCombatWith(iOwner);
```
Comment (ccrs) [inline src/server/game/Combat/ThreatManager.cpp]: I would've kept this (addthreat) in Unit just because of these setincombatwithcalls, threatmanager shouldnt use any unit methods right?

Comment (Treeston) [inline src/server/game/Combat/ThreatManager.cpp]: This is one of the compat layer things. In the full PR, this is handled properly.

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp [hunk 4]
```diff
@@ -200,6 +200,49 @@ void ScriptedAI::DoPlaySoundToSet(WorldObject* source, uint32 soundId)
...
+    if (!who)
+        who = me;
+    who->GetThreatManager().ResetThreat(victim);
+}
+
+void ScriptedAI::ResetThreatList(Unit* who)
+{
+    if (!who)
+        who = me;
+    who->ResetThreatList();
```
Comment (Shauren) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: what about this one?

Comment (Treeston) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: Missed it.

File: src/common/Utilities/Containers.h [hunk 5]
```diff
@@ -87,8 +87,9 @@ namespace Trinity
         template<class C>
         inline auto SelectRandomContainerElement(C const& container) -> typename std::add_const<decltype(*std::begin(container))>::type&
         {
+            ASSERT(Size(container));
             auto it = std::begin(container);
-            std::advance(it, urand(0, uint32(Size(container)) - 1));
+            std::advance(it, urand(0, Size(container) - 1));
```
Comment (Shauren) [inline src/common/Utilities/Containers.h]: W4 level warning, dont remove the cast

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 6]
```diff
@@ -405,8 +408,3 @@ bool FarthestTargetSelector::operator()(Unit const* target) const
 
     return true;
 }
-
-void SortByDistanceTo(Unit* reference, std::list<Unit*>& targets)
```
Comment (Shauren) [inline src/server/game/AI/CoreAI/UnitAI.cpp]: This function was added to remove the need to include Object in UnitAI.h, don't remove it.

File: src/server/game/AI/CoreAI/UnitAI.h [hunk 7]
```diff
@@ -152,94 +152,174 @@ class TC_GAME_API UnitAI
...
+            // ignore the first <offset> elements
+            while (offset)
+            {
+                targetList.pop_front();
+                --offset;
+            }
+            // and filter by predicate
+            for (auto it = targetList.begin(), end = targetList.end(); it != end;)
+            {
+                if (predicate(*it))
```
Comment (Shauren) [inline src/server/game/AI/CoreAI/UnitAI.h]: This is a change that can break all existing scripts - don't swap the order of operations, filter THEN apply offset

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/AI/CoreAI/UnitAI.cpp (hunk 6 above)
Comment (Treeston) [inline src/server/game/AI/CoreAI/UnitAI.cpp]: fixed

File: src/server/game/AI/CoreAI/UnitAI.h (hunk 7 above)
Comment (Treeston) [inline src/server/game/AI/CoreAI/UnitAI.h]: If anything, this _fixes_ a bunch of existing scripts. This behavior is what (most) script devs assumed.

Comment (Shauren) [inline src/server/game/AI/CoreAI/UnitAI.h]: Honestly, if your goal is to get this merged as fast as possible then you should revert this order change (reviewing it would require checking every call to SelectTarget in scripts, something I don't want to do in the scope of this PR)

Comment (Treeston) [inline src/server/game/AI/CoreAI/UnitAI.h]: Done, will make it a separate PR.

File: src/server/game/AI/CoreAI/UnitAI.h [hunk 8]
```diff
@@ -152,94 +152,175 @@ class TC_GAME_API UnitAI
...
             {
-                if (currentVictim != nullptr && hostileRef->getTarget() != currentVictim && predicate(hostileRef->getTarget()))
-                    targetList.push_back(hostileRef->getTarget());
-                else if (currentVictim == nullptr && predicate(hostileRef->getTarget()))
-                    targetList.push_back(hostileRef->getTarget());
+                targetList.pop_front();
+                --offset;
+            }
+            // and finally filter by predicate
+            for (auto it = targetList.begin(), end = targetList.end(); it != end;)
```
Comment (Shauren) [inline src/server/game/AI/CoreAI/UnitAI.h]: my previous filter then offset comment also applies to single unit selection, not just list

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/AI/CoreAI/UnitAI.h (hunk 8 above)
Comment (Treeston) [inline src/server/game/AI/CoreAI/UnitAI.h]: right, obviously.

:x:

**Review by Shauren — CHANGES_REQUESTED** [review]
It crashes.

**Comment by Shauren** [comment]
- Teleport into Icecrown Citadel
- .mod hp 200000000
- .gm off
- pull everything from the start to Lord Marrowgar
- .gm on

And it crashes.
https://gist.github.com/Shauren/a25f026fd5b6edd014a81c4ea0719716

**Comment by Treeston** [comment]
Missing `nullptr` check in one of the compatibility layer methods, fixed.
